students/GuyCChan/session07/html_render.py

```diff
@@ -18,22 +19,46 @@ def __init__(self, content=None, **kwargs):
     def append(self, new_content):
         self.contents.append(new_content)
 
-    def render(self, out_file):
-        # loop through the list of contents (see func above)
-        out_file.write("<{}>\n".format(self.tag))
+    def _open_close_tags(self):
+        """ Made open and close tags for all elements."""
+        attrs = " ".join(["{}='{}'".format(key, value) for key, value in self.attributes.items()])
+        # I really couldn't figure out comletely how to word these lines, particularly the .strip(),
+        # and had to resort to plagiarising! But now I think I begin to appreciate how it's done.
+        if attrs.strip():
```

Comment:
this could be a touch cleaner with:

```
        if self.attributes:
            attrs = " ".join(["{}='{}'".format(key, value) for key, value in self.attributes.items()])
            open_tag = "<{} {}>".format(self.tag, attrs)
        else:
            open_tag = "<{}>".format(self.tag)
```

but this works, too.